### ISSUE TITLE
ReleaseClub Contract

### ISSUE THREAD

**Issue by neatonk** [ReleaseClub/contracts#1]
Moving our discussion from Discord to here for posterity:

@dhruv035 and I discussed creating a minimal smart contract to capture the state of each release club on chain.

- - -

### Currently, each release club will need to store:
- [x] the owner address
- [x] a list of member addresses
- [x] a list of releases with contract and token id (see [comment](https://github.com/ReleaseClub/release-club-contracts/issues/1#issuecomment-1206720891)).
- [ ] the club name, up to 20 bytes: revert if longer than 20 bytes.

### Each member should able to:
- [x] add a batch of releases
- [ ] add a batch of members

### The owner should be able to:
- [x] do anything a member can
- [x] transfer ownership (and become a member?)
- [x] set the club name
- [x] add a batch of members

### Anyone should be able to:
- [x] get the club name
- [x] get the owner address
- [x] get the list of member addresses
- [x] get the list of releases

### Each release must:
- [ ] be on a contract owned by the sender (The owner MAY NOT add releases for other members).
- [ ] be an ERC721 (1155's are not supported by Zora's indexer yet). See #10.

**Comment by neatonk** [comment]
## ERC721 use cases

These use cases all use the same spec with different semantics, which will affect the way we wish to display them and how we represent them in the ReleaseClub contract. Not sure if we can detect that a contract represents a single release on-chain via the provided contract address so we may have to let the user decide (e.g. choose single/multi release when adding the contract). 

- single-release, single-user ( e.g. editions created on create.zora.co)
- multi-release, single-user (e.g. a custom contract)
- multi-release, multi-user (e.g. mintsongs, and catalog)

**Comment by neatonk** [comment]
I'd also like to consider what a contract upgrade would look like in the future.

Some options I can think of:
- non-upgradable - deploy a new contract and switch to it. Existing clubs are stuck unless we provide migration path. Would be costly to recreate a club on a new contract.
- [UUPS or transparent upgradable proxy](https://docs.openzeppelin.com/upgrades-plugins/1.x/#proxy-patterns): more work up front, but has many benefits.

**Comment by neatonk** [comment]
I think batching can be done with OZ MultiCall plugin

**Comment by neatonk** [comment]
The enumerate priveledged users by role ...
```ts
const minterCount = await myToken.getRoleMemberCount(MINTER_ROLE);

const members = [];
for (let i = 0; i < minterCount; ++i) {
    members.push(await myToken.getRoleMember(MINTER_ROLE, i));
}
```
See https://docs.openzeppelin.com/contracts/3.x/access-control#querying-privileged-accounts

**Comment by josealonso** [comment]
What is the member identifier, a number ?

**Comment by josealonso** [comment]
Just started to write a Typescript interface to make sure I understand these specs. This interface will be used by the frontend, @salieflewis.

**Comment by josealonso** [comment]
Written the first version of the contract on the new *contracts* repo, @neatonk

**Comment by josealonso** [comment]
`The club name, up to 20 bytes: revert if longer than 20 bytes`. This feature is implemented in this PR #13
